packages/validator/src/processor.ts

```diff
@@ -0,0 +1,204 @@
+  let tmpResult: Result = { valid: true, messages: [], output: {} };
+  let tmpRoots: Roots = { specs: specs, nonRedundantParams: nonRedundantParams, output: {} };
+
+  for (const key of Object.keys(template)) {
+    if (key === '__ignore') {
+      for (const copy of Object.keys(specs)) {
+        if (!nonRedundantParams[copy]) {
+          nonRedundantParams[copy] = specs[copy];
+        }
+      }
+
+      break;
+    }
+
+    switch (key) {
+      case '__conditions':
+        for (const condition of template[key]) {
+          messages.push(...validateCondition(specs, condition, nonRedundantParams, paramPath, roots, paramPathPrefix));
+        }
+
+        break;
+
+      case '__regexp':
+        messages.push(...validateRegexp(specs, template, paramPath));
+        break;
+
+      case '__keyRegexp':
+        messages.push(...validateRegexp(specs, template, paramPath, true));
+        break;
+
+      case '__maxSize':
+        if (template[key] < specs.length) {
+          messages.push(logger.error(`${paramPath} must contain ${template[key]} or less items`));
+        }
+
+        break;
+
+      // validate array
+      case '__arrayItem':
+        // nonRedundantParams has to have the same structure as template
+        if (!nonRedundantParams) {
+          nonRedundantParams = [];
+        }
+
+        // validate each item in specs
+        for (let i = 0; i < specs.length; i++) {
+          nonRedundantParams.push({});
+          const result = processSpecs(
+            specs[i],
+            template[key],
+            `${paramPath}[${i}]`,
+            nonRedundantParams[i],
+            roots,
+            paramPathPrefix
+          );
+          messages.push(...result.messages);
+        }
+
+        break;
+
+      // in specs can be any parameter, should validate all of them according to whats in the template
+      case '__objectItem':
+        for (const item of Object.keys(specs)) {
+          // insert empty type of item into nonRedundantParams
+          nonRedundantParams[item] = utils.getEmptyNonRedundantParam(item, template, nonRedundantParams, specs[item]);
+
+          const result = processSpecs(
+            specs[item],
+            template[key],
+            `${paramPath}${paramPath ? '.' : ''}${item}`,
+            nonRedundantParams[item],
+            roots,
+            paramPathPrefix
+          );
+          messages.push(...result.messages);
+        }
+
+        break;
+
+      case '__optional':
+        messages.push(...validateOptional(specs, template[key], paramPath, nonRedundantParams, roots, paramPathPrefix));
+
+        break;
+
+      // determines level of message for single parameter, currently used only in regexp validation
+      case '__level':
+        break;
+
+      case '__any':
+        if (!isAnyParamValid(specs, template[key], paramPath, nonRedundantParams, roots)) {
+          messages.push(logger.error(`Required conditions not met in ${paramPath}`));
+        }
```

Comment:
For consistency, since all of your other validation subroutines return a message, I think this one should too; i.e. change the name to `validateAny` and push the output of that function to `messages`, so that all of your `validate...` subroutines have similar inputs/outputs.

Reply:
Changed in 87d7be6